account-kit/signer/src/base.ts

```diff
@@ -809,6 +817,74 @@ export abstract class BaseAlchemySigner<TClient extends BaseSignerClient>
     });
   };
 
+  /**
+   * This method lets you adapt your AlchemySigner to a viem WalletClient.
+   *
+   * @param {ToViemWalletClientParams} opts - options
+   * @example
+   * ```ts
+   * import { AlchemyWebSigner } from "@account-kit/signer";
+   *
+   * const signer = new AlchemyWebSigner({
+   *  client: {
+   *    connection: {
+   *      rpcUrl: "/api/rpc",
+   *    },
+   *    iframeConfig: {
+   *      iframeContainerId: "alchemy-signer-iframe-container",
+   *    },
+   *  },
+   * });
+   *
+   * const client = signer.toViemWalletClient();
+   * ```
+   *
+   * @throws if your signer is not authenticated
+   * @returns {WalletClient} a viem WalletClient object that can be used with viem's wallet client
+   */
+  toViemWalletClient = <
+    TTransport extends Transport,
+    TChain extends Chain | undefined = undefined,
+    TRpcSchema extends RpcSchema | undefined = undefined
+  >({
+    transport,
+    ...opts
+  }: Omit<
+    WalletClientConfig<TTransport, TChain, LocalAccount, TRpcSchema>,
+    "account"
+  >): WalletClient<TTransport, TChain, LocalAccount, TRpcSchema> => {
+    const viemAccount = this.toViemAccount();
+
+    // Construct fallback transport: https://github.com/wevm/viem/blob/59d2c093fe435e97684d34266f72c432dd6a0b3a/src/clients/createClient.ts#L231-L234
+    const fallbackTransport = transport({
+      chain: opts.chain,
+      pollingInterval: opts.pollingInterval ?? 4_000,
+    });
+
+    return createWalletClient({
+      ...opts,
+      account: viemAccount,
+      transport: custom({
+        async request({ method, params }) {
+          switch (method) {
+            case "personal_sign":
+              return viemAccount.signMessage(
+                isHex(params[0])
+                  ? { message: { raw: params[0] } }
+                  : { message: params[0] }
```

Comment:
It should be technically possible to sign ascii/utf characters that represent hex characters, and this conversion prevents that by re-interpreting the characters to their raw value if they happen to be hex characters.

I think viem defaults to having the parameters to `personal_sign` be either a string or an object containing a field called `raw` - could we continue to use that type here?

Reply:
That's what the check here is already doing no? Passing raw if hex so that it doesn't get treated as string

Reply:
No, Adam's saying that this makes it impossible to sign the string "0x1234" because it will get interpreted as bytes `0x12`, `0x34`. (Or not _impossible_, but someone would need to pass the hex encoding of the above string, i.e. "0x307831323334"). Adam's saying that the caller should be responsible for specifying whether their string should be interpreted as hex rather than us trying to infer it.

But actually I think it's wrong to accept the type with `raw` here too. This method is trying to emulate the `personal_sign` method of the eth-jsonrpc API, and therefore should _always_ interpret the input has hex ([Metamask's docs](https://docs.metamask.io/wallet/reference/json-rpc-methods/personal_sign/)). Getting non-hex input here should be an error.

Reply:
Ah, good catch. So the real fix here is to just throw an error instead of handling the case of `{ message: params[0] }`, right?

Reply:
Yeah, I think so.